File: java/iroha_android/src/main/java/org/hyperledger/iroha/android/model/instructions/MergeRwasInstruction.java
package org.hyperledger.iroha.android.model.instructions;

import java.util.Collections;
import java.util.LinkedHashMap;
import java.util.Map;
import java.util.Objects;

/**
 * Typed builder for the {@code MergeRwas} instruction.
 *
 * <p>The payload stores the canonical JSON representation of the merge request.
 */
public final class MergeRwasInstruction implements InstructionTemplate {

  private static final String ACTION = "MergeRwas";

  private final String mergeJson;
  private final Map<String, String> arguments;

  private MergeRwasInstruction(final Builder builder) {
    this(builder, builder.canonicalArguments());
  }

  private MergeRwasInstruction(final Builder builder, final Map<String, String> argumentOrder) {
    this.mergeJson = builder.mergeJson;
    this.arguments =
        Collections.unmodifiableMap(new LinkedHashMap<>(Objects.requireNonNull(argumentOrder)));
  }

  /** Returns the canonical JSON representation of the merge request payload. */
  public String mergeJson() {
    return mergeJson;
  }

  @Override
  public InstructionKind kind() {
    return InstructionKind.CUSTOM;
  }

  @Override
  public Map<String, String> toArguments() {
    return arguments;
  }

  public static MergeRwasInstruction fromArguments(final Map<String, String> arguments) {
    final Builder builder = builder().setMergeJson(require(arguments, "merge"));
    return new MergeRwasInstruction(builder, new LinkedHashMap<>(arguments));
  }

  private static String require(final Map<String, String> arguments, final String key) {
    final String value = arguments.get(key);
    if (value == null || value.isBlank()) {
      throw new IllegalArgumentException("Instruction argument '" + key + "' is required");
    }
    return value;
  }

  public static Builder builder() {
    return new Builder();
  }

  @Override
  public boolean equals(final Object obj) {
    if (this == obj) {
      return true;
    }
    if (!(obj instanceof MergeRwasInstruction other)) {
      return false;
    }
    return Objects.equals(mergeJson, other.mergeJson);
  }

  @Override
  public int hashCode() {
    return Objects.hash(mergeJson);
  }

  public static final class Builder {
    private String mergeJson;

    private Builder() {}

    public Builder setMergeJson(final String mergeJson) {
      this.mergeJson = Objects.requireNonNull(mergeJson, "mergeJson");
      return this;
    }

    public MergeRwasInstruction build() {
      if (mergeJson == null || mergeJson.isBlank()) {
        throw new IllegalStateException("mergeJson must be provided");
      }
      return new MergeRwasInstruction(this);
    }

    private Map<String, String> canonicalArguments() {
      final Map<String, String> args = new LinkedHashMap<>();
      args.put("action", ACTION);
      args.put("merge", mergeJson);
      return args;
    }
  }
}

File: .taira-source-c80bf4d85a97/kotlin/core-jvm/src/main/java/org/hyperledger/iroha/sdk/core/model/instructions/MergeRwasInstruction.kt
package org.hyperledger.iroha.sdk.core.model.instructions

private const val ACTION = "MergeRwas"

/**
 * Typed representation of the `MergeRwas` instruction.
 *
 * The payload stores the canonical JSON representation of the merge request.
 */
class MergeRwasInstruction private constructor(
    /** Returns the canonical JSON representation of the `MergeRwas` payload. */
    @JvmField val mergeJson: String,
    override val arguments: Map<String, String>,
) : InstructionTemplate {

    override val kind: InstructionKind = InstructionKind.CUSTOM

    constructor(mergeJson: String) : this(
        mergeJson = validated(mergeJson, "mergeJson"),
        arguments = linkedMapOf(
            "action" to ACTION,
            "merge" to mergeJson,
        ),
    )

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is MergeRwasInstruction) return false
        return mergeJson == other.mergeJson
    }

    override fun hashCode(): Int = mergeJson.hashCode()

    companion object {
        @JvmStatic
        fun fromArguments(arguments: Map<String, String>): MergeRwasInstruction {
            val mergeJson = require(arguments, "merge")
            return MergeRwasInstruction(
                mergeJson = mergeJson,
                arguments = LinkedHashMap(arguments),
            )
        }

        private fun require(arguments: Map<String, String>, key: String): String {
            val value = arguments[key]
            require(!value.isNullOrBlank()) { "Instruction argument '$key' is required" }
            return value
        }

        private fun validated(value: String, name: String): String {
            require(value.isNotBlank()) { "$name must not be blank" }
            return value
        }
    }
}
